### PR TITLE
Add FeeCollectorHooks to Distribution Module

### PR DESCRIPTION
**Pull request by wr1159**
## Add FeeCollectorHooks to Distribution Module

Related to https://github.com/Switcheo/carbon/pull/951

Distribution Module already has their own hooks but there is no set function for hooks so I created a FeeCollectorHook that will be used in the their own hook.

---

### CODE REVIEW

File: x/distribution/keeper/keeper.go [hunk 1]
```diff
@@ -38,7 +39,7 @@ func NewKeeper(
 	cdc codec.BinaryCodec, storeService store.KVStoreService,
 	ak types.AccountKeeper, bk types.BankKeeper, sk types.StakingKeeper,
 	feeCollectorName, authority string,
-) Keeper {
+) *Keeper {
```
Comment (yan-soon) [inline x/distribution/keeper/keeper.go:42]: why is this changed to a pointer? is it bc of carbon `app.go` related dependency init?

File: x/distribution/types/expected_keepers.go [hunk 2]
```diff
@@ -60,3 +60,7 @@ type StakingHooks interface {
 	AfterValidatorCreated(ctx context.Context, valAddr sdk.ValAddress) error // Must be called when a validator is created
 	AfterDelegationModified(ctx context.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) error
 }
+
+type DistributionHook interface {
+	BeforeFeeCollectorSend(ctx context.Context, feeCollector sdk.ModuleAccountI) error
+}
```
Comment (yan-soon) [inline x/distribution/types/expected_keepers.go:66]: i think this can just be in `types/hooks.go`, since `expected_keepers` is mainly used for other keeper interfaces

File: x/distribution/keeper/keeper.go (hunk 1 above)
Comment (wr1159) [reply x/distribution/keeper/keeper.go:42]: Yeah and JR wanted the assigning of distr hooks to be done with pointers

File: x/distribution/types/expected_keepers.go (hunk 2 above)
Comment (wr1159) [reply x/distribution/types/expected_keepers.go:66]: okay will change next monday if not urgent

File: x/distribution/keeper/allocation.go [hunk 3]
```diff
@@ -19,11 +19,15 @@ func (k Keeper) AllocateTokens(ctx context.Context, totalPreviousPower int64, bo
 	// called in BeginBlock, collected fees will be from the previous block
 	// (and distributed to the previous proposer)
 	feeCollector := k.authKeeper.GetModuleAccount(ctx, k.feeCollectorName)
+	err := k.hooks.BeforeFeeCollectorSend(ctx, feeCollector)
```
Comment (RavenXce) [inline x/distribution/keeper/allocation.go:22]: i thought we no longer pass the feecollector address

**Review by RavenXce — CHANGES_REQUESTED** [review]

File: x/distribution/types/expected_keepers.go (hunk 2 above)
Comment (wr1159) [reply x/distribution/types/expected_keepers.go]: Changed in 1a8bde7e2bba1c1e299eab286282dcb6afbe376c also

**Review by yan-soon — APPROVED** [review]
